Creation and setup of taf repositories enhancements

renatav opened the pull request:
## Description (e.g. "Related to ...", etc.)

- Added new commands for:
  - setting up a new yubikey
  - exporting a public key from inserted yubikey
  - adding a new signing key (currently assumes that all keys are stored on yubikeys)

- Improved creation of repositories:
  - check keystore before inserting all of yubikeys
  - created standardized yubikey prompt
  -  keep retrying to access the yubikey rather than failing out with an exception (can take a few seconds for computer to recognize yubikey after inserted)
  -  store pins globallyso don't have to keep typing password.
  - made the process more robust
  - added support for mixing keys stored on yubikeys with keys from keystore files

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/taf/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/

danixeee commented on taf/yubikey.py:
```diff
@@ -19,12 +20,25 @@
+
+
+def add_key_pin(serial_num, pin):
+    global _pins_dict
```
No need for `global` here.

danixeee commented on taf/yubikey.py:
```diff
@@ -296,3 +310,101 @@ def setup(
+            pin_repeat,
+            prompt_message,
+        )
+        if success:
```
Should we raise exception if `success` is `False` or if not, we should `return None, None`?

renatav replied:
The idea is to repeat `_read_and_check_yubikey` until success is `True`.

danixeee commented on taf/repository_tool.py:
```diff
@@ -335,6 +370,14 @@ def add_targets(self, data, targets_role="targets", files_to_keep=None):
             if target_path.is_file():
                 self._add_target(targets_obj, str(target_path), previous_custom)
 
+    def add_yubikey_external_signature_provider(self, role, name):
```
Not used anymore

danixeee commented on taf/exceptions.py:
```diff
@@ -39,6 +43,10 @@ def __init__(self, message):
         super().__init__("root", message)
 
 
+class KeystoreError(TAFError):
```
Does it make sense to subclass `YubikeyError`?

renatav replied:
That seems confusing to me, but it would make sense to create a common ancestor class. Something like `KeyError`, sublassed by `KeystoreError` and `YubikeyError`?

danixeee replied:
I guess we can live it like this then

renatav replied:
If at some point we need to catch both exceptions, we should consider creating a superclass. At the moment, that does not occur anywhere

danixeee commented on taf/yubikey.py:
```diff
@@ -296,3 +310,101 @@ def setup(
+            raise InvalidPINError("No retries left. YubiKey locked.")
+        if not valid_pin:
+            if not click.confirm(
+                f"Incorrect PIN. Do you want to try again? {retries} retires left."
```
`retires` -> `retries`

danixeee commented on taf/developer_tool.py:
```diff
@@ -45,6 +45,127 @@
+                if role_key_infos is not None and role in role_key_infos:
+                    is_yubikey = role_key_infos[role].get("yubikey")
+                if is_yubikey is None:
+                    is_yubikey = click.confirm(f"Load {key_name} from YubiKey?")
```
I got confused when I saw this question. Should we ask something like: "Do you want to sing {key_name} with Yubikey?" ?

danixeee replied:
`loaded_yubikeys` is passed as a wrong argument. `yk.yubikey_prompt(key_name, role, taf_repo, loaded_yubikeys=loaded_yubikeys)`

danixeee commented on taf/developer_tool.py:
```diff
@@ -45,6 +45,127 @@
+                    yk.yubikey_prompt(key_name, role, taf_repo, False, loaded_yubikeys)
+                else:
+                    pem = getpass(
+                        f"Enter {key_name} private key without its header and footer"
```
Could you add a new line to the end of a message?

danixeee commented on taf/developer_tool.py:
```diff
@@ -45,6 +45,127 @@
+                    pem = getpass(
+                        f"Enter {key_name} private key without its header and footer"
+                    )
+                    pem = _form_private_pem(pem)
```
Does it make sense to loop until a valid PEM is entered? If I enter something like "abcde" I get following error `securesystemslib.exceptions.CryptoError: RSA (public, private) tuple cannot be generated from the encrypted PEM string: Could not deserialize key data.`

renatav replied:
Moved all this code to a new function where we loop until a valid PEM is entered and check if it's a valid key of the given role

danixeee commented on taf/developer_tool.py:
```diff
@@ -155,76 +276,52 @@ def build_auth_repo(
-            str(Path(keystore, f"{key_name}.pub")), scheme
+        public_key = _read_public_key_from_keystore(keystore, key_name)
+        private_key = _read_private_key_from_keystore(
+            keystore, key_name, roles_key_info, key_num
```
`scheme` should be passed when loading a key

danixeee commented on taf/developer_tool.py:
```diff
@@ -155,76 +276,52 @@ def build_auth_repo(
-
     # if it does not, generate the keys and print the output
     else:
         key = generate_rsa_key()
```
`scheme` should be passed when generating a key

danixeee commented on taf/developer_tool.py:
```diff
@@ -547,6 +733,37 @@ def _check_key_id(expected_key_id):
+    _setup_new_yubikey(serial_num, certs_dir)
+
+
+def _setup_new_yubikey(serial_num, certs_dir=None, scheme=DEFAULT_RSA_SIGNATURE_SCHEME):
```
`key` should be returned from this function

danixeee commented on taf/developer_tool.py:
```diff
@@ -155,76 +276,52 @@ def build_auth_repo(
+        registered = True
+        # set Yubikey expiration date
+        # add_yubikey_external_signature_provider
+        role_obj.add_verification_key(key, expires=YUBIKEY_EXPIRATION_DATE)
```
`key` is `None` if I choose to setup a new yubikey

renatav replied:
There was a return missing in `setup_new_yubikey`.

danixeee commented on taf/developer_tool.py:
```diff
@@ -547,6 +733,37 @@ def _check_key_id(expected_key_id):
+    pub_key_pem = yk.setup(pin, cert_cn, cert_exp_days=EXPIRATION_INTERVAL).decode(
+        "utf-8"
+    )
+    scheme = DEFAULT_RSA_SIGNATURE_SCHEME
```
This line should be removed

danixeee commented on taf/developer_tool.py:
```diff
@@ -547,6 +733,37 @@ def _check_key_id(expected_key_id):
+    )
+    scheme = DEFAULT_RSA_SIGNATURE_SCHEME
+    key = import_rsakey_from_pem(pub_key_pem, scheme)
+    if certs_dir is None:
```
Could we move lines 758-766 to a new function, so we can export certificate to `certs_dir` when using existing yubikeys?

renatav replied:
Done

renatav commented:
@danixeee I think that I addressed all comments. Tested all variations - new yubikeys, existing yubikeys, loading keys from keystore and entering keys

danixeee commented:
@renatav When I follow the same flow for testing (https://github.com/openlawlibrary/taf/pull/79#pullrequestreview-321621916) I cannot sign targets with the following command `taf sign_targets --repo-path ./law --keystore ./data/keystore/` and the reason is that _targets_ key is on my yubikey and not in a keystore (and cannot be exported). I think we should slightly update the function to use keystore for keys that are not generated on yubikeys (that is specified in _keys.json_) and not all keys. Does this make sense or `--keystore` flag means that _all_ keys will be loaded from files?
Other than that, everything seems fine and if this comment does not make sense, we can merge this.
